File: cpp/203/Remove_Linked_List_Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode *t = new ListNode(0);
        t->next = head;
        ListNode *pre = t;
        while(head){
            if(head->val==val){
                pre->next = head->next;
                delete head;
                head = pre->next;
            }
            else{
                pre = head;
                head = head->next;
            }
        }
        head = t->next;
        delete t;
        return head;
    }
};

File: python/203/Remove_Linked_List_Elements.py
class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        if not head:
            return head
        newhead = ListNode(0)
        newhead.next = head
        pre, p = newhead, head
        while p:
            if p.val != val:
                pre = pre.next
                p = p.next
            else:
                pre.next = p.next
                p = pre.next
        return newhead.next
